tests/components/derivative/test_sensor.py

```diff
@@ -97,132 +70,88 @@
-
-    # Testing a energy sensor with non-monotonic intervals and values
-    for time, value in [(20, 10), (30, -10)]:
-        now = dt_util.utcnow() + timedelta(seconds=time)
-        with patch("homeassistant.util.dt.utcnow", return_value=now):
-            hass.states.async_set(entity_id, value, {}, force_update=True)
-            await hass.async_block_till_done()
-
-    state = hass.states.get("sensor.power")
-    assert state is not None
-
-    assert round(float(state.state), config["sensor"]["round"]) == -2
+    await setup_tests(
+        hass, {"unit_time": "s"}, times=[20, 30], values=[10, -10], expected_state=-2
+    )
 
 
 async def test_dataSet6(hass):
     """Test derivative sensor state."""
-    config = {
-        "sensor": {
-            "platform": "derivative",
-            "name": "power",
-            "source": "sensor.energy",
-            "round": 2,
-        }
-    }
-
-    assert await async_setup_component(hass, "sensor", config)
+    await setup_tests(hass, {}, times=[0, 60], values=[0, 1 / 60], expected_state=1)
 
-    entity_id = config["sensor"]["source"]
-    hass.states.async_set(entity_id, 0, {})
-    await hass.async_block_till_done()
 
-    # Testing a energy sensor with non-monotonic intervals and values
-    for time, value in [(20, 0), (30, 36000)]:
+async def test_data_moving_average_for_discrete_sensor(hass):
+    """Test derivative sensor state."""
+    # We simulate the following situation:
+    # The temperature rises 1 °C per minute for 1 hour long.
+    # There is a data point every second, however, the sensor returns
+    # the temperature rounded down to an integer value.
+    # We use a time window of 10 minutes and therefore we can expect
+    # (because the true derivative is 1 °C/min) an error of less than 10%.
+
+    temperature_values = []
+    for temperature in range(60):
+        temperature_values += [temperature] * 60
+    time_window = 600
+
+    times = list(range(len(temperature_values)))
+    config, entity_id = await _setup_sensor(
+        hass, {"time_window": {"seconds": time_window}, "unit_time": "min", "round": 1}
+    )  # two minute window
+
+    for time, value in zip(times, temperature_values):
         now = dt_util.utcnow() + timedelta(seconds=time)
         with patch("homeassistant.util.dt.utcnow", return_value=now):
             hass.states.async_set(entity_id, value, {}, force_update=True)
             await hass.async_block_till_done()
 
-    state = hass.states.get("sensor.power")
-    assert state is not None
-
-    assert round(float(state.state), config["sensor"]["round"]) == 1
+        if time_window < time < times[-1] - time_window:
+            state = hass.states.get("sensor.power")
+            derivative = round(float(state.state), config["sensor"]["round"])
+            # Test that the error is never more than
+            # (time_window_in_minutes / true_derivative * 100) = 10%
+            assert abs(1 - derivative) <= 0.1
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=4938379)